torch/distributed/optim/optimizer.py

```diff
@@ -0,0 +1,90 @@
+import torch.distributed.rpc as rpc
+import torch.distributed.autograd as dist_autograd
+
+from collections import defaultdict
+from threading import Lock
+
+
+class _LocalOptimizer:
+    def __init__(self, optim_cls, local_params_rref, *args, **kwargs):
+        self.optim = optim_cls(
+            [rref.local_value().wait() for rref in local_params_rref],
+            *args,
+            **kwargs)
+        self.lock = Lock()
```

Comment:
This is the lock mentioned [here](https://github.com/pytorch/pytorch/pull/28910#discussion_r341277257) to, right? How would we modify this to get Hogwild?

Reply:
Yes. There's no simple way of modifying the current implementation to get hogwild. We'd have to either 1) change the interface (FunctionalOptimizer); or some alternative such as keeping an object pool / or thread local instances of optim.Optimizers to avoid gradient sharing across threads. I can write a comment on why the lock is there.

Reply:
Cool, just wanted to verify. Thanks!

Reply:
Wouldn't removing this lock give us some form of hogwild? I'm not sure how accurate this example is: https://pytorch.org/docs/stable/notes/multiprocessing.html#hogwild, although it does something similar where accumulating gradients on the tensors and running the optimizer could interleave among different processes.